### PR TITLE
Don't modify original query object during preprocessing

### PR DESCRIPTION
**Pull request by hsyuan**
For the following madlib query `select madlib.svec_plus('{1}:{4}', '{1}:{5}');`, we first do const folding before passing query to Orca. But during function evaluation, the 2 arguments are modified by madlib function. Before this patch, even the query object is copied in query_tree_mutator(), but it is flat copy. It just copy Datum of constvalue, not the actual data pointing by the constvalue. So to avoid assertion failure when folding constants, we have to do deep copy the whole query object.

### CODE REVIEW

File: src/backend/optimizer/util/clauses.c [hunk 1]
```diff
@@ -2100,9 +2101,10 @@ fold_constants(PlannerGlobal *glob, Query *q, ParamListInfo boundParams, Size ma
...
 	context.max_size = max_size;
 	
+	query = (Query *) copyObject(q);
 	return (Query *) query_or_expression_tree_mutator
 						(
-						(Node *) q,
+						(Node *) query,
 						eval_const_expressions_mutator,
 						&context,
 						0
```
Comment (dhanashreek89) [inline src/backend/optimizer/util/clauses.c]: `query_tree_mutator()` on L4596 creates a copy of the Query node and that is what we want; is there any way we can pass appropriate flags such that `query_tree_mutator()` will create a copy for us before modifying it ?

Comment (hsyuan) [reply src/backend/optimizer/util/clauses.c]: I can't find what you are pointing. The query is modified when processing expression, not the query tree. The function is called `query_or_expression_tree_mutator`.

Comment (dhanashreek89) [reply src/backend/optimizer/util/clauses.c]: `query_or_expression_tree_mutator()` uses `query_tree_mutator()` eventually  : 
```
Node *
query_or_expression_tree_mutator(Node *node,
								 Node *(*mutator) (),
								 void *context,
								 int flags)
{
	if (node && IsA(node, Query))
		return (Node *) query_tree_mutator((Query *) node,
										   mutator,
										   context,
										   flags);
       ....
}
```
or am I missing something ?

Comment (hsyuan) [reply src/backend/optimizer/util/clauses.c]: Ah, I see what you mean. But by default, it will copy the query object if we give the flag 0.

Comment (hsyuan) [reply src/backend/optimizer/util/clauses.c]: I think the difference is that in the query tree mutator just flat copy the query node, what does this patch do is deep copy the query tree. But yeah, we can give the flag `QTW_DONT_COPY_QUERY` to disable double copy the query tree.

**Comment by vraghavan78** [comment]
Can you add a test?

**Comment by dhanashreek89** [comment]
LGTM. Could you please add more description such as how this assert got tripped and a test ?

**Comment by hsyuan** [comment]
I want to add a test too, but need to come up with a smaller and portable test to repro the MADlib issue. It is not feasible to simply put a MADlib test into ICG.

**Comment by hsyuan** [comment]
Heikki, in tracker story 152048829, madlib team reported there is an assertion failure in `DETAIL:  FailedAssertion("!(equal(qcopy, query) && "Preprocessing should not modify original query object")", File: "transform.c", Line: 60)`. 

After debugging, I found that the madlib UDF modified the input argument. That is why it has assertion failure. There is nothing wrong with GPDB code, so I will close this PR.

**Comment by hsyuan** [comment]
Well, it turns out we still needs this patch. Both madlib and postgis has function that modified input argument in-place.

For the following madlib query `select madlib.svec_plus('{1}:{4}', '{1}:{5}');`, we first do const folding before passing query to Orca. But during function evaluation, the 2 arguments are modified by madlib function. Before this patch, even the query object is copied in `query_tree_mutator()`, but it is flat copy. It just copy Datum of `constvalue`, not the actual data pointing by the `constvalue`. So to avoid assertion failure when folding constants, we have to do deep copy the whole query object.

**Comment by hlinnaka** [comment]
On 26 October 2017 23:23:42 CEST, Haisheng Yuan <notifications@github.com> wrote:
>Well, it turns out we still needs this patch. Both madlib and postgis
>has function that modified input argument in-place.

Sounds like a bug in madlib and postgis.


- Heikki

**Comment by hsyuan** [comment]
Do you mean madlib and postgis function should not modify the datum data of the input arguments?

**Comment by hlinnaka** [comment]
On 27 October 2017 07:54:01 CEST, Haisheng Yuan <notifications@github.com> wrote:
>Do you mean madlib and postgis function should not modify the datum
>data of the input arguments?

Yes.


- Heikki

**Comment by hlinnaka** [comment]
See the warning at https://www.postgresql.org/docs/current/static/xfunc-c.html#xfunc-c-basetype.

**Comment by hsyuan** [comment]
Thanks @hlinnaka, I will close this PR and let madlib team know.

**Pull request closed without merging**